File: sdk/sdk-java/src/main/java/dev/httpmarco/polocloud/sdk/java/player/PlayerProvider.java
package dev.httpmarco.polocloud.sdk.java.player;

import dev.httpmarco.polocloud.common.future.FutureConverterKt;
import dev.httpmarco.polocloud.shared.player.PolocloudPlayer;
import dev.httpmarco.polocloud.shared.player.SharedPlayerProvider;
import dev.httpmarco.polocloud.shared.service.Service;
import dev.httpmarco.polocloud.v1.player.*;
import io.grpc.ManagedChannel;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;
import java.util.concurrent.CompletableFuture;

public class PlayerProvider implements SharedPlayerProvider<PolocloudPlayer> {

    private final PlayerControllerGrpc.PlayerControllerBlockingStub blockingStub;
    private final PlayerControllerGrpc.PlayerControllerFutureStub futureStub;

    public PlayerProvider(ManagedChannel channel) {
        this.blockingStub = PlayerControllerGrpc.newBlockingStub(channel);
        this.futureStub = PlayerControllerGrpc.newFutureStub(channel);
    }

    @Override
    public @NotNull List<PolocloudPlayer> findAll() {
        return this.blockingStub.findAll(PlayerFindRequest.getDefaultInstance()).getPlayersList().stream().map(PolocloudPlayer.Companion::bindSnapshot).toList();
    }

    @Override
    public @NotNull CompletableFuture<List<PolocloudPlayer>> findAllAsync() {
        return FutureConverterKt.completableFromGuava(this.futureStub.findAll(PlayerFindRequest.newBuilder().build()), findAllPlayerResponse -> findAllPlayerResponse.getPlayersList().stream().map(PolocloudPlayer.Companion::bindSnapshot).toList());
    }

    @Override
    @Nullable
    public PolocloudPlayer findByName(@NotNull String name) {
        return this.blockingStub.findByName(PlayerFindByNameRequest.newBuilder().setName(name).build()).getPlayersList().stream().map(PolocloudPlayer.Companion::bindSnapshot).findFirst().orElse(null);
    }

    @Override
    @NotNull
    public CompletableFuture<PolocloudPlayer> findByNameAsync(@NotNull String name) {
        return FutureConverterKt.completableFromGuava(this.futureStub.findByName(PlayerFindByNameRequest.newBuilder().setName(name).build()),
                findGroupResponse -> findGroupResponse.getPlayersList().stream().map(PolocloudPlayer.Companion::bindSnapshot).findFirst().orElse(null));
    }

    @Override
    @NotNull
    public List<PolocloudPlayer> findByService(@NotNull String serviceName) {
        return this.blockingStub.findByService(PlayerFindByServiceRequest.newBuilder().setCurrentServiceName(serviceName).build()).getPlayersList().stream().map(PolocloudPlayer.Companion::bindSnapshot).toList();
    }

    @Override
    @NotNull
    public CompletableFuture<List<PolocloudPlayer>> findByServiceAsync(@NotNull Service service) {
        return FutureConverterKt.completableFromGuava(this.futureStub.findByService(PlayerFindByServiceRequest.newBuilder().setCurrentServiceName(service.name()).build()),
                findByServiceRequest -> findByServiceRequest.getPlayersList().stream().map(PolocloudPlayer.Companion::bindSnapshot).toList());
    }

    @Override
    public int playerCount() {
        return this.blockingStub
                .playerCount(PlayerCountRequest.getDefaultInstance())
                .getCount();
    }
}

File: sdk/sdk-rust/src/providers/player_provider.rs
use crate::polocloud::player_controller_client::PlayerControllerClient;
use crate::polocloud::{PlayerFindByNameRequest, PlayerFindByServiceRequest, PlayerFindRequest};
use crate::structs::PolocloudPlayer;
use tonic::transport::Channel;
use tonic::{Request, Status};

pub struct PlayerProvider {
    player_stub: PlayerControllerClient<Channel>,
}
impl PlayerProvider {
    pub fn new(channel: Channel) -> Self {
        Self {
            player_stub: PlayerControllerClient::new(channel),
        }
    }

    pub async fn find_all_async(&mut self) -> Result<Vec<PolocloudPlayer>, Status> {
        let request = Request::new(PlayerFindRequest::default());
        let response = self.player_stub.find_all(request).await?;

        let players = response.into_inner().players
            .into_iter()
            .map(|player_snapshot| PolocloudPlayer::from_snapshot(player_snapshot))
            .collect();

        Ok(players)
    }

    pub async fn find_by_name_async(&mut self, name: String) -> Result<Option<PolocloudPlayer>, Status> {
        let request = Request::new(PlayerFindByNameRequest {name});
        let response = self.player_stub.find_by_name(request).await?;

        let player_option = response.into_inner().players
            .into_iter()
            .map(|player_snapshot| PolocloudPlayer::from_snapshot(player_snapshot)).nth(0);

        Ok(player_option)
    }

    pub async fn find_by_service_async(&mut self, service_name: String) -> Result<Vec<PolocloudPlayer>, Status> {
        let request = Request::new(PlayerFindByServiceRequest {current_service_name: service_name});
        let response = self.player_stub.find_by_service(request).await?;

        let players = response.into_inner().players
            .into_iter()
            .map(|player_snapshot| PolocloudPlayer::from_snapshot(player_snapshot))
            .collect();

        Ok(players)
    }
}

